src/Workspaces/Core/Portable/Editing/ImportAdderService.cs

```diff
@@ -18,44 +20,240 @@ namespace Microsoft.CodeAnalysis.Editing
+            var nodes = root.DescendantNodesAndSelf().Where(isInSpan);
+            var (importDirectivesToAdd, context) = strategy switch
+            {
+                Strategy.AddImportsFromSymbolAnnotations
+                    => GetImportDirectivesFromAnnotatedNodesAsync(nodes, root, model, addImportsService, generator, cancellationToken),
+                Strategy.AddImportsFromSyntaxes
+                    => GetImportDirectivesFromSyntaxesAsync(nodes, ref root, model, addImportsService, generator, cancellationToken),
+                _ => throw new InvalidEnumArgumentException(nameof(strategy), (int)strategy, typeof(Strategy)),
+            };
 
-            var namespacesToAdd = new HashSet<INamespaceSymbol>();
-            namespacesToAdd.AddRange(nodesWithExplicitNamespaces.Select(
-                n => GetExplicitNamespaceSymbol(n, model)));
+            var placeSystemNamespaceFirst = options.GetOption(GenerationOptions.PlaceSystemNamespaceFirst, document.Project.Language);
 
-            var generator = SyntaxGenerator.GetGenerator(document);
-            var imports = namespacesToAdd.Select(ns => generator.NamespaceImportDeclaration(ns.ToDisplayString()).WithAdditionalAnnotations(Simplifier.Annotation))
-                                         .ToArray();
+            root = addImportsService.AddImports(model.Compilation, root, context, importDirectivesToAdd, placeSystemNamespaceFirst);
 
-            // annotate these nodes so they get simplified later
-            var newRoot = root.ReplaceNodes(
-                nodesWithExplicitNamespaces,
-                (o, r) => r.WithAdditionalAnnotations(Simplifier.Annotation));
+            return document.WithSyntaxRoot(root);
+        }
 
-            var placeSystemNamespaceFirst = options.GetOption(GenerationOptions.PlaceSystemNamespaceFirst, document.Project.Language);
-            var addImportsService = document.GetLanguageService<IAddImportsService>();
-            var finalRoot = addImportsService.AddImports(
-                model.Compilation, newRoot, newRoot, imports, placeSystemNamespaceFirst);
+        protected abstract INamespaceSymbol GetExplicitNamespaceSymbol(SyntaxNode node, SemanticModel model);
 
-            return document.WithSyntaxRoot(finalRoot);
+        /// <summary>
+        /// Gets the namespace <paramref name="node"/> is contained in, or null otherwise
+        /// </summary>
+        protected abstract INamespaceSymbol GetContainedNamespace(SyntaxNode node, SemanticModel model);
+
+        private SyntaxNode GenerateNamespaceImportDeclaration(INamespaceSymbol namespaceSymbol, SyntaxGenerator generator)
+        {
+            return generator
+                .NamespaceImportDeclaration(namespaceSymbol.ToDisplayString(SymbolDisplayFormats.NameFormat))
+                .WithAdditionalAnnotations(Simplifier.Annotation, Formatter.Annotation);
```

Comment:
Note: my recollection is that when it comes to imports we prefer to not simplify them.  That's because users want to be able to look at the `using X.Y.Z` and understand waht X.Y.Z is without having to go "wait... waht does X resolve off of from a higher up using".

Reply:
I think that it is necessary to simplify them, as import annotations can sometimes be added unnecessarily, and we need to be able to remove them.

Reply:
Ok.  I'm ok with the entire 'import/using' being removed.  We just have to make sure we don't *simplify* the 'name' on the RHS of the import/using.  If this is only about the former (which we should doc), then this is fine :)